include/deal.II/matrix_free/fe_evaluation.h

```diff
@@ -1505,22 +1604,177 @@ class FEEvaluationAccess<1,1,Number> : public FEEvaluationBase<1,1,Number>
+ * different DoFHandler objects for velocity and pressure, the second argument
+ * to the FEEvaluation objects specify the respective component 0 for velocity
+ * and 1 for pressure. For further examples of vector-valued problems, the
+ * deal.II test suite includes a few additional examples as well, e.g. the
+ * Stokes operator described above is found at
+ * https://github.com/dealii/dealii/blob/master/tests/matrix_free/matrix_vector_stokes_noflux.cc
+ *
+ * <h3>Handling several integration tasks and data storage</h3>
+ *
+ * The design of FEEvaluation and MatrixFree separates the geometry from the
+ * basis functions. Therefore, several DoFHandler objects (or the same
+ * DoFHandler equipped with different constraint objects) can share the same
+ * geometry information like in the Stokes example above. All gemoetry is
+ * cached once in MatrixFree, so FEEvaluation does not need to do expensive
+ * initialization calls and rather sets a few pointers. This realization is
+ * based on the idea that the geometry information is needed only once also
+ * when several fields are evaluated, in a departure from FEValues which sets
+ * up the internal mapping data for each field. If for example a
+ * multi-component PDE involves the shape values on one component and the
+ * shape gradient on the other, no efficiency is lost if both are based on the
+ * same MatrixFree object where the update flags specify that both @p
+ * update_values , @p update_gradients , and @p update_jxw_values are
+ * given. The selection of desired quantities of shape values is through the
+ * flags in the evaluate() or integrate calls and the access at quadrature
+ * points:
+ *
+ * @code
+ * phi1.evaluate(true, false);
+ * phi2.evaluate(false, true);
+ * for (unsigned int q_index=0; q_index<phi1.n_q_points; ++q_index)
+ *   {
+ *     VectorizedArray<double> val1 = phi1.get_value(q);
+ *     Tensor<1,dim,VectorizedArray<double> > grad2 = phi2.get_gradient(q);
+ *     Point<dim,VectorizedArray<double> > point = phi1.get_quadrature_point(q);
+ *     // ... some complicated formula combining those three...
+ *   }
+ * @endcode
+ *
+ * In the loop over quadrature points, one can ask any of the two FEEvaluation
+ * objects mdash; it does not really matter which one because they only keep
+ * pointers to the quadrature point data mdash; to provide the quadrature
+ * point location.
+ *
+ * This observation also translates to the case when different differential
+ * operators are implemented in a program, for example the action of a mass
+ * matrix in one matrix-vector product and the action of a stiffness matrix in
+ * another one. Only a single MatrixFree object is necessary, maintaining full
+ * efficiency if the different FEEvaluation objects in the two local calls. In
+ * other words, a user does not need to bother about being conservative when
+ * providing update_flags to the initialization of MatrixFree for efficiency
```

Comment:
could you please reference this discussion from `MatrixFree` class documentation?